pkg/sfu/downtrack.go

```diff
@@ -798,615 +610,833 @@ func (d *DownTrack) UpdateStats(packetLen uint32) {
+
+func (f *Forwarder) SetMaxSpatialLayer(spatialLayer int32) bool {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if spatialLayer == f.maxSpatialLayer {
+		return false
+	}
+
+	f.maxSpatialLayer = spatialLayer
+	return true
+}
+
+func (f *Forwarder) MaxSpatialLayer() int32 {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	return f.maxSpatialLayer
+}
+
+func (f *Forwarder) CurrentSpatialLayer() int32 {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	return f.currentSpatialLayer
+}
+
+func (f *Forwarder) TargetSpatialLayer() int32 {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	return f.targetSpatialLayer
+}
+
+func (f *Forwarder) SetMaxTemporalLayer(temporalLayer int32) bool {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if temporalLayer == f.maxTemporalLayer {
+		return false
+	}
+
+	f.maxTemporalLayer = temporalLayer
+	return true
+}
+
+func (f *Forwarder) MaxTemporalLayer() int32 {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	return f.maxTemporalLayer
+}
+
+func (f *Forwarder) MaxLayers() (int32, int32) {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	return f.maxSpatialLayer, f.maxTemporalLayer
+}
+
+func (f *Forwarder) GetForwardingStatus() ForwardingStatus {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	if f.targetSpatialLayer == InvalidSpatialLayer {
+		return ForwardingStatusOff
+	}
+
+	if f.targetSpatialLayer < f.maxSpatialLayer {
+		return ForwardingStatusPartial
+	}
+
+	return ForwardingStatusOptimal
+}
+
+func (f *Forwarder) AdjustAllocation(availableChannelCapacity uint64, brs [3][4]uint64) (isPausing, isResuming bool, bandwidthRequested, optimalBandwidthNeeded uint64) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	isPausing = false
+	isResuming = false
+	bandwidthRequested = 0
+	optimalBandwidthNeeded = 0
+
+	if f.kind == webrtc.RTPCodecTypeAudio || f.muted {
+		return
+	}
+
+	optimalBandwidthNeeded = uint64(0)
+	// LK-TODO for temporal preference, traverse the bitrates array the other way
+	for i := f.maxSpatialLayer; i >= 0; i-- {
+		for j := f.maxTemporalLayer; j >= 0; j-- {
+			if brs[i][j] == 0 {
+				continue
+			}
+			if optimalBandwidthNeeded == 0 {
+				optimalBandwidthNeeded = brs[i][j]
+			}
+			if brs[i][j] < availableChannelCapacity {
+				isResuming = f.targetSpatialLayer == InvalidSpatialLayer
+				bandwidthRequested = brs[i][j]
+
+				f.targetSpatialLayer = int32(i)
+				f.targetTemporalLayer = int32(j)
+				return
+			}
 		}
+	}
 
-		err = d.writeRTPHeaderExtensions(&hdr)
-		if err != nil {
-			return err
+	if optimalBandwidthNeeded != 0 {
+		// no layer fits in the available channel capacity, disable the track
+		isPausing = f.targetSpatialLayer != InvalidSpatialLayer
+
+		f.currentSpatialLayer = InvalidSpatialLayer
+		f.targetSpatialLayer = InvalidSpatialLayer
+
+		f.currentTemporalLayer = InvalidTemporalLayer
+		f.targetTemporalLayer = InvalidTemporalLayer
+	}
+	return
+}
+
+func (f *Forwarder) IncreaseAllocation(brs [3][4]uint64) (increased bool, bandwidthRequested, optimalBandwidthNeeded uint64) {
+	increased = false
```

Comment:
golang will init variable to zero values.

Reply:
Thank you @cnderrauber . Will make a small commit to clean these up. Thank you for the review.

Reply:
Addressed in this commit - https://github.com/livekit/livekit-server/pull/213/commits/04d3b2c683fa0798a44f65de53cd0aaa233075d9. @davidzhao Going to merge this, but please do review when you get a chance and let me know of any improvements.